sql/database.go

```diff
@@ -442,6 +693,167 @@ func (db *sqliteDatabase) Close() error {
+		stmt.BindText(1, toPath)
+	}, nil); err != nil {
+		return fmt.Errorf("vacuum into %s: %w", toPath, err)
+	}
+	return nil
+}
+
+// copyMigrateDB performs a copy-based migration of the database.
+// The source database is always closed by this function.
+// Upon success, the migrated database is opened.
+func (db *sqliteDatabase) copyMigrateDB(config *conf) (finalDB *sqliteDatabase, err error) {
+	defer db.Close()
+
+	dbPath, migratedPath, err := dbMigrationPaths(config.uri)
+	if err != nil {
+		return nil, err
+	}
+	if migratedPath == "" {
+		return nil, fmt.Errorf("cannot migrate database, only file DBs are supported: %s", config.uri)
+	}
+
+	// Instead of just copying the source database to the temporary migration DB, use VACUUM INTO.
+	// This is somewhat slower but achieves two goals:
+	// 1. The lock is held on the source database while it's being copied
+	// 2. If the source database has a lot of free pages for whatever reason, those
+	// are not copied, saving disk space
+	config.logger.Info("making a temporary copy of the database",
+		zap.String("path", dbPath),
+		zap.String("target", migratedPath))
+	if err := db.vacuumInto(migratedPath); err != nil {
+		return nil, errors.Join(err, deleteDB(migratedPath))
+	}
+
+	// Opening the temporary migrated DB runs the actual migrations on it.
+	// We disable vacuuming here because we're going to vacuum the temporary DB
+	// into the original one.
+	opts := []Opt{
+		WithLogger(config.logger),
+		WithConnections(1),
+		WithTemp(),
+		WithDatabaseSchema(config.schema),
+	}
+	if config.ignoreSchemaDrift {
+		opts = append(opts, WithIgnoreSchemaDrift())
+	}
+	migratedDB, err := Open("file:"+migratedPath, opts...)
+	if err != nil {
+		return nil, errors.Join(
+			fmt.Errorf("process temporary DB %s: %w", migratedPath, err),
+			deleteDB(migratedPath))
+	}
+	tempDBReady := false
+	defer func() {
+		err = errors.Join(err, migratedDB.Close())
+		if !tempDBReady {
+			err = errors.Join(err, deleteDB(migratedPath))
+		}
+	}()
+
+	// Make sure the temporary DB is fully synced to the disk before creating the marker file.
+	// We don't need wal_checkpoint(TRUNCATE) here as we're going to delete the temporary DB.
+	if _, err := migratedDB.Exec("PRAGMA wal_checkpoint(FULL)", nil, nil); err != nil {
+		return nil, fmt.Errorf("checkpoint temporary DB %s: %w", migratedPath, err)
+	}
+
+	// Create the marker file to indicate that the migration is complete.
+	// Make sure the file is written to the disk before closing the database.
+	// We could create a table in the temporary database instead of the marker file,
+	// but as the temporary database is opened without PRAGMA journal_mode=OFF
+	// and PRAGMA synchronous=OFF, it may become corrupt in case of a crash or power
+	// outage, so we avoid trying to open it.
+	markerPath := migratedPath + "_done"
+	if f, err := os.Create(markerPath); err != nil {
+		return nil, fmt.Errorf("create marker file %s_done: %w", migratedPath, err)
+	} else {
```

Comment:
We should try to avoid nesting if we can:

```suggestion
	f, err := os.Create(markerPath)
	if err != nil {
		return nil, fmt.Errorf("create marker file %s_done: %w", migratedPath, err)
	}
```

and not putting everything else into the `else` block.

Reply:
Fixed